url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/mvd/XMLGuideFile.java

```diff
@@ -0,0 +1,390 @@
+			Version v = mvd.versions.get( i );
+			VersionInfo w = new VersionInfo( v.group, v.backup, 
+				new String(v.shortName), new String(v.longName), "" );
+			versions.add( w );
+		}
+	}
+	/**
+	 * Get the version information for a given version
+	 * @param vId the version id
+	 * @return the version info
+	 */
+	public VersionInfo getVersionInfo( short vId ) throws MVDToolException
+	{
+		if ( vId-1 < 0 || vId-1 > versions.size() )
+			throw new MVDToolException("Invalid version id "+vId );
+		VersionInfo vi = versions.get( vId-1 );
+		return vi;
+	}
+	/**
+	 * Get the number of versions stored in this archive
+	 * @return the number of versions
+	 */
+	public int numVersions()
+	{
+		return versions.size();
+	}
+	/**
+	 * Get the name of a group
+	 * @param gId the group id
+	 * @return the group name
+	 */
+	public String getGroupName( short gId ) throws MVDToolException
+	{
+		if ( gId-1 < 0 || gId-1 > groups.size() )
+			throw new MVDToolException("Invalid group id "+gId );
+		Group g = groups.get( gId-1 );
+		return g.name;
+	}
+	/**
+	 * Set the file name for the version just written out
+	 * @param vId the id of the version
+	 * @param file the name of the file
+	 */
+	public void setVersionFile( short vId, String file ) 
+		throws MVDToolException
+	{
+		if ( vId-1 < 0 || vId-1 > versions.size() )
+			throw new MVDToolException("Invalid version id "+vId );
+		VersionInfo vi = versions.get( vId-1 );
+		vi.setFile( file );
+	}
+	/**
+	 * Get the names of all versions in the archive
+	 * @return an array of version file names in the archive folder
+	 */
+	public String[] getVersionFileNames()
+	{
+		String[] files = new String[versions.size()];
+		for ( int i=0;i<files.length;i++ )
+		{
+			VersionInfo vi = versions.get( i );
+			files[i] = vi.file;
+		}
+		return files;
+	}
+	/**
+	 * Get the description of the MVD
+	 * @return a String
+	 */
+	public String getDescription()
+	{
+		return description;
+	}
+	/**
+	 * Write an MVDGuideFile out as XML.
+	 * @param dst the destination guide XML file
+	 * @throws an exception if it couldn't write the file or find a 
+	 * serialiser
+	 */
+	public void externalise( File dst ) throws MVDException
+	{
+		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
+		try 
+		{
+			DocumentBuilder db = dbf.newDocumentBuilder();
+			// create an instance of DOM
+			Document dom = db.newDocument();
+			Element rootElement = dom.createElement( MVDGUIDE );
+			Element descElement = dom.createElement( DESCRIPTION );
+			descElement.setTextContent( description );
+			dom.appendChild( rootElement );
+			rootElement.appendChild( descElement );
+			writeGroups( dom, rootElement );
+			writeVersions( dom, rootElement );
+			// print it out manually since DOM serialisation escapes CRLF  
+			// as &13;LF on the Mac and this behaviour can't be turned off. 
+			// We need a portable externalise routine that works the same 
+			// on all platforms
+			FileOutputStream fos = new FileOutputStream( dst );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *fos*. The resource is referenced at lines 239, 240. The resource is closed at line 241. There are other execution paths that do not contain closure statements, for example, when XMLPretty.writeNode throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).